### ISSUE TITLE
Internal Files Creation Settings

### ISSUE THREAD

**Issue by realmonster** [ladislav-zezula/StormLib#96]
### In short
There is no way to set up flags, date, locale, and compression to intertal files: (signature),  (listfile), (attributes).
### What for?
Old game patches may have issues with opening _modern_ internal files. Thus, giving errors of corrupt file.
### Current implementation
Normaly there are two ways to replace/create file with this library:
- **SFileAddFileEx** which is load file from file system.
- **SFileCreateFile** which create MPQFile writeable handler, and allows most customization.
None of them permits to write internal files. It's intended hardcoded restriction.
Both of them using **SFileAddFile_Init** to create file. This function is internal one, and does not have such restriction.

You can try to overwrite internal file with **SFileAddFile_Init**, which is kinda pain, but even if you did it, anyways it will be overwritten again by **SFileFlushArchive** called from **SFileCloseArchive**.

### Default settings
- **Flags** = **GetDefaultSpecialFileFlags(...)** | **MPQ_FILE_REPLACEEXISTING**
- **Date** = **N/A** (0)
- **LANG_NEUTRAL**
- **Compression** = **ZLib**

### Workarounds
1. Simple: modify SFileAttributes.cpp, SFileListFile.cpp, SFileVerify.cpp with hardcoded values that you need.
2. Dirty hack: before closing file:
 - call SFileFlushArchive, it will generate updated internal files.
 - read existing internal files into memory, because we erase them next
 - call InvalidateInternalFiles(mpq)
 - mpq->dwFlags |= MPQ_FLAG_SAVING_TABLES. This is to disable auto InvalidateInternalFiles when new file is made
 - call DefragmentFileTable(mpq)
 - write internal files using **SFileAddFile_Init**
 - remove from mpq->dwFlags flags NEW and NONE for appropriate internal list files.
May be I'm overcomplicating dirty hack, but now I'm doing in this way. Perhaps you can just replace files and clear flags. Not sure.

### Compression
There is **SFileSetDataCompression** function which should set default compression.
It turns out... its value is used only in **SFileAddFileEx**.
But internal files using **SFileAddFile_Init**. Thus, it has no effect for internal files.
They **always** use **ZLib** compression. In my case, it's not acceptable.

### Ways to implement it
1. Use default compression for saving internal files.
2. Remember old properties of internal files during invalidation. (does not fix compression, because it's not property, there is only flag "I'm compressed")
3. Introduce new function for setting things up. (messy)
4. Add toggle to allow rewrite internal files, and disable their automatic overwriting.

I suggest (1) and (4) at the same time. Also, I can implement it and make pull request.
Writing things here, just to start discussion.

**Comment by ladislav-zezula** [comment]
When opening existing archive:
- TMPQArchive::dwFileFlags1 gets flags for (listfile)
- TMPQArchive::dwFileFlags2 gets flags for (attributes)
- TMPQArchive::dwFileFlags3 gets flags for (signature)

The intent is to save those special files with the same flags like they were before, because like you wrote, various games have various flags for tose files.

Then creating new archive: the SFileCreateArchive2, takes the SFILE_CREATE_MPQ structure as parameter. Its dwFileFlags1, dwFileFlags2 and dwFileFlags3 are used as input for those internal files.

**Comment by realmonster** [comment]
So it's keeping flags. Good.
But if I wan't to change them, I need to create other MPQ. Bad.

I'll repeat. In my case, game crash because ZLib applied for internal files.
And SFileCreateArchive2 won't help.

**Comment by realmonster** [comment]
Actually, SFileCreateArchive2 does help in my case!
Because I can create MPQ without compression of internal files, and thus, avoid them being ZLib compressed.
Other workaround is doing this way:
```cpp
((TMPQArchive*)mpq)->dwFileFlags1 &= ~MPQ_FILE_COMPRESS;
((TMPQArchive*)mpq)->dwFileFlags2 &= ~MPQ_FILE_COMPRESS;
((TMPQArchive*)mpq)->dwFileFlags3 &= ~MPQ_FILE_COMPRESS;
```
Confirmed working.

But there is one corner case, if resulting dwFileFlags1 exactly MPQ_FILE_EXISTS, then two lines:
https://github.com/ladislav-zezula/StormLib/blob/e9f9b9bc58c63a72285920b889c8d434e74e8a28/src/SFileListFile.cpp#L389
https://github.com/ladislav-zezula/StormLib/blob/e9f9b9bc58c63a72285920b889c8d434e74e8a28/src/SBaseCommon.cpp#L342
Will give compressed file.

So, in my top post, **I was wrong about Default flags: if internal file was in archive, it will keep same flags**.

**Comment by ladislav-zezula** [comment]
Fixed in commit adac4aa0f586c7aa2a3c19aa4e687d827dc86195.

You can now set MPQ_FILE_DEFAULT_INTERNAL (a new constant, equals to 0xFFFFFFFF) as dwFileFlags1, dwFileFlags2 and dwFileFlags3. StormLib will set default flags if you do.

If you want to set just MPQ_FILE_EXISTS, then you can and it will stay unchanged.
